src/libcore/num/mod.rs

```diff
@@ -2037,6 +2037,113 @@ assert_eq!((-a).rem_euclid(-b), 1);
             }
         }
 
+        doc_comment! {
+            concat!("Returns the logarithm of the number with respect to an arbitrary base.
+
+Returns `None` if the number is lower than 1.
+
+This method may not be optimized owing to implementation details;
+`self.log2()` can produce results more efficiently for base 2, and
+`self.log10()` can produce results more efficiently for base 10.
+
+# Examples
+
+```
+#![feature(int_log)]
+
+let five = 5", stringify!($SelfT), ";
+
+// log5(5) == 1
+let result = five.log(5);
+
+assert_eq!(result, Some(1));
+```"),
+            #[unstable(feature = "int_log", issue = "70887")]
+            #[must_use = "this returns the result of the operation, \
+                          without modifying the original"]
+            #[inline]
+            pub fn log(self, base: Self) -> Option<Self> {
+                let logb2 = |x: Self| ((1 as Self).leading_zeros() - x.leading_zeros()) as Self;
+
+                if self <= 0 || base <= 1 {
+                    None
+                } else {
+                    let mut n = 0;
+                    let mut r = self;
+
+                    // Optimization for 128 bit wide integers.
+                    if mem::size_of::<Self>() * 8 == 128 {
+                        let b = logb2(self) / (logb2(base) + 1);
+                        n += b;
+                        r /= base.pow(b as u32);
+                    }
+
+                    while r >= base {
+                        r /= base;
+                        n += 1;
+                    }
+                    Some(n)
+                }
+            }
+        }
+
+        doc_comment! {
+            concat!("Returns the base 2 logarithm of the number.
+
+Returns `None` if the number is lower than 1.
+
+# Examples
+
+```
+#![feature(int_log)]
+
+let two = 2", stringify!($SelfT), ";
+
+// log2(2) == 1
+let result = two.log2();
+
+assert_eq!(result, Some(1));
+```"),
+            #[unstable(feature = "int_log", issue = "70887")]
+            #[must_use = "this returns the result of the operation, \
+                          without modifying the original"]
+            #[inline]
+            pub fn log2(self) -> Option<Self> {
+                if self <= 0 {
+                    None
+                } else {
+                    let log = (1 as Self).leading_zeros() - self.leading_zeros();
```

Comment:
This should really be using `intrinsics::ctlz_nonzero`, as doing so results in substantially-better ASM for the default target, as even slightly better ASM for newer CPUs: https://rust.godbolt.org/z/b9Jqi_

Reply:
Oh cool!